os/arch/arm/src/imxrt/imxrt_hyperflash.c

```diff
@@ -649,6 +834,20 @@ ssize_t up_progmem_write(size_t addr, const void *buf, size_t count)
 
 	return count;
 }
+#else
+/************************************************************************************
+ * Name: up_flashinitialize
+ *
+ * Description:
+ *   Return an initialize MTD device instance.  MTD devices are not registered
+ *   in the file system, but are created as instances that can be bound to
+ *   other functions (such as a block or character driver front end).
+ *
+ ************************************************************************************/
+FAR struct mtd_dev_s *up_flashinitialize(void)
```

Comment:
mtd driver should override up_flashinitialze.
Hyperflash should override up_spiinialize

During boot phase, board should init its sub devices.
```
spi_dev_s *spi= up_spiinitialize(portnum); //CONFIG needed for each port... 
mtd_dev_s *mtd = up_flashinitialize(spi);
```

up_spiinitialize should be overrided in board specific code(c.f. imxrt_spi.c), it should init spi related code.
In your case, your support 2 type of spi, spi & flexspi(quad, double quad spi). 
```
FAR struct spi_dev_s *up_spiinitialize(int bus) {
   ...
   switch(port) {
     case CONFIG_IMXRT_SPI_PORT_0 : //if it is flexspi, initialze it
        init_flexspi;
        init_gpio_pin(PORT_FOR_THIS);
     ....
}
```

Reply:
As I said to you, iMXRT1050 has independent IP Block named FlexSPI and different from LPSPI.
Can you tell me why I should implement the SPI interface in detail?
I have already reviewed your mtd source code.
I cannot find the source code dependent on SPI interface in mtd driver.

Reply:
You already understand hyperflash can use spi, dual spi, quad spi, dual quad spi(Flexspi).
So you have to implement common mtd layer, please do not make dependecy with flex spi only. You have to use common spi interface.

MTD driver has to request I/O operation through the spi bus.
In other words, MTD driver request this i/o operation through the spi_ops_s. (See MACRO in spi.h)
Below is calling process.
MTD -> SPI_MACRO -> FLASH Driver(implemented spi_ops_s) -> FLASH Driver's priv apis(flexSPI or lpspi)

To sum up, 
1. your flash driver should implement below
```
struct spi_dev_s;
struct spi_ops_s {
#ifndef CONFIG_SPI_OWNBUS
	int (*lock)(FAR struct spi_dev_s *dev, bool lock);
#endif
	void (*select)(FAR struct spi_dev_s *dev, enum spi_dev_e devid, bool selected);
	uint32_t (*setfrequency)(FAR struct spi_dev_s *dev, uint32_t frequency);
	void (*setmode)(FAR struct spi_dev_s *dev, enum spi_mode_e mode);
	void (*setbits)(FAR struct spi_dev_s *dev, int nbits);
	uint8_t (*status)(FAR struct spi_dev_s *dev, enum spi_dev_e devid);
#ifdef CONFIG_SPI_CMDDATA
	int (*cmddata)(FAR struct spi_dev_s *dev, enum spi_dev_e devid, bool cmd);
#endif
	uint16_t (*send)(FAR struct spi_dev_s *dev, uint16_t wd);
#ifdef CONFIG_SPI_EXCHANGE
	void (*exchange)(FAR struct spi_dev_s *dev, FAR const void *txbuffer, FAR void *rxbuffer, size_t nwords);
#else
	void (*sndblock)(FAR struct spi_dev_s *dev, FAR const void *buffer, size_t nwords);
	void (*recvblock)(FAR struct spi_dev_s *dev, FAR void *buffer, size_t nwords);
#endif
	int (*registercallback)(FAR struct spi_dev_s *dev, spi_mediachange_t callback, void *arg);
};
```
2. You already have a  flash driver's priv functions, flexSPI.
    When your spi layer called by mtd, you can call any of priv functions, flexSPI

3. Please assume that hyperflash & your nor flash added on the flexSPI.
   In this case you have to maintain both of 2 flexSPI like a your gpio pins.

4. Please See functions in m25px.c, below is an example
```
static void m25p_writeenable(struct m25p_dev_s *priv)
{
	/* Select this FLASH part */

	SPI_SELECT(priv->dev, SPIDEV_FLASH, true);

	/* Send "Write Enable (WREN)" command */

	(void)SPI_SEND(priv->dev, M25P_WREN);

	/* Deselect the FLASH */

	SPI_SELECT(priv->dev, SPIDEV_FLASH, false);
	fvdbg("Enabled\n");
}
```
 Here we use SPI Macro to call function which implemented spi_ops_s interface.
 suppose that m25px is on your board(lpspi), SPI_SEND will call below
```
static uint16_t imxrt_lpspi_send(FAR struct spi_dev_s *dev, uint16_t wd)
{
	FAR struct imxrt_lpspidev_s *priv = (FAR struct imxrt_lpspidev_s *)dev;
	uint32_t regval;
	uint16_t ret;

	DEBUGASSERT(priv && priv->spibase);

	imxrt_lpspi_writeword(priv, (uint32_t) wd);

	while ((imxrt_lpspi_getreg32(priv, IMXRT_LPSPI_SR_OFFSET) & LPSPI_SR_RDF(1U)) !=
		LPSPI_SR_RDF(1U));

	ret = imxrt_lpspi_readword(priv);

	/* Check and clear any error flags (Reading from the SR clears the error
	 * flags).
	 */

	regval = imxrt_lpspi_getreg32(priv, IMXRT_LPSPI_SR_OFFSET);

	spiinfo("Sent: %04x Return: %04x Status: %02x\n", wd, ret, regval);

	UNUSED(regval);
	return ret;
}
```

Reply:
So to appy what I mentioned in above, MTD should know pointer of each of spi dev's pointer.
in imxrt_lpspi, it already implemented up_spiinitialize.
You have to use this spi dev's pointer when you initialize mtd driver.(please see m25p_initialize)